x-pack/legacy/plugins/integrations_manager/server/integrations.ts

```diff
@@ -0,0 +1,231 @@
+  const paths = await Registry.getArchiveInfo(`${pkgkey}.tar.gz`);
+  const toBeSavedObjects = new Map();
+
+  for (const path of paths) {
+    collectReferences(path, toBeSavedObjects, 'dashboard');
+  }
+
+  return Array.from(toBeSavedObjects.values(), ensureJsonValues);
+}
+
+function getAsset(key: string) {
+  const value = cacheGet(key);
+  if (value !== undefined) {
+    const json = value.toString('utf8');
+    return JSON.parse(json);
+  }
+}
+
+function collectReferences(
+  path: string,
+  toBeSavedObjects: Map<string, SavedObject> = new Map(),
+  desiredType: string = 'dashboard'
+) {
+  const [pkgkey, service, type, file] = path.split('/');
+  if (type !== desiredType) return;
+  if (toBeSavedObjects.has(path)) return;
+  if (!/\.json$/.test(path)) return;
+
+  const asset = getAsset(path);
+  if (!asset.type) asset.type = type;
+  if (!asset.id) asset.id = file.replace('.json', '');
+  toBeSavedObjects.set(path, asset);
+
+  for (const reference of asset.references) {
+    collectReferences(
+      `${pkgkey}/${service}/${reference.type}/${reference.id}.json`,
+      toBeSavedObjects,
+      reference.type
+    );
+  }
+}
+
+// the assets from the registry are malformed
+// https://github.com/elastic/integrations-registry/issues/42
+function ensureJsonValues(obj: SavedObject) {
+  const { attributes } = obj;
+  if (
+    attributes.kibanaSavedObjectMeta &&
+    typeof attributes.kibanaSavedObjectMeta.searchSourceJSON !== 'string'
+  ) {
+    attributes.kibanaSavedObjectMeta.searchSourceJSON = JSON.stringify(
+      attributes.kibanaSavedObjectMeta.searchSourceJSON
+    );
+  }
+  ['optionsJSON', 'panelsJSON', 'uiStateJSON', 'visState']
+    .filter(key => typeof attributes[key] !== 'string')
+    .forEach(key => (attributes[key] = JSON.stringify(attributes[key])));
+  return obj;
+}
+
+function createIntegrationList(
+  registryItems: RegistryList,
+  integrationObjects: InstallationSavedObject[]
+) {
+  const integrationList: IntegrationList = [];
+
+  for (const item of registryItems) {
+    const installedObject: InstallationSavedObject | undefined = integrationObjects.find(
+      ({ id }) => id === `${item.name}-${item.version}`
+    );
+
+    const integration = createInstallationObject(item, installedObject);
+    integrationList.push(integration);
+  }
+
+  return integrationList.sort(sortByName);
+}
+
+function createInstallationObject(
+  item: RegistryPackage,
+  savedObject?: InstallationSavedObject
+): IntegrationInfo;
+function createInstallationObject(
+  item: RegistryListItem,
+  savedObject?: InstallationSavedObject
+): IntegrationListItem;
+function createInstallationObject(
+  obj: RegistryPackage | RegistryListItem,
+  savedObject?: InstallationSavedObject
+) {
+  return savedObject
+    ? {
+        ...obj,
+        status: 'installed',
+        savedObject,
+      }
+    : {
+        ...obj,
+        status: 'not_installed',
+      };
+}
```

Comment:
I’m aiming to have the API (HTTP, JS, etc) types specified in `common/types` so that client and server, or others, can share the typing for docs, validation, etc. Like the work in https://github.com/elastic/kibana/pull/40041. I just learned that Beats is doing something similar https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/beats_management/common/return_types.ts and heard they worked really well.

IMO, the client and server should consume/implement the spec; the spec shouldn’t fall out from their return values. If the response shape changes that should be explicitly changed in the interface/type in `common`.

Reply:
Yeah, in APM we did the same thing around defining common/shared return types, but used a `ReturnType` to guarantee that the client consumed exactly what the server returned: https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/apm/server/lib/transactions/get_top_transactions/index.ts#L23-L25

As long as the contract is secure it probably doesn't matter too much, but it would be nice to at least make clear which types are "Response" types with some kind of naming convention, perhaps?

Reply:
Sure. Names are definitely in flux.

Can we address those in later PRs as they come up vs iterating in this one? It’s got a lot of other useful stuff I’d like to land.

Reply:
I’ll log an issue later tonight. I’m worried about letting the confusion grow and become worse.

Reply:
Filed https://github.com/elastic/kibana/issues/40558 around naming

I want to keep types in common, distinct from client/seever

Reply:
@jfsiii just to be clear, you mean the types for the HTTP API responses, right? Not all types for the project? Just clarifying -- makes sense to me to keep the API response types there.

Reply:
Not all types, no. Just those that are common/shared between `public` and `server` or other top-level directories.